src/Symfony/Component/HttpKernel/Event/ControllerEvent.php

```diff
@@ -43,6 +44,32 @@ public function getController(): callable
 
     public function setController(callable $controller): void
     {
+        if (isset($this->controller) ? $controller !== $this->controller : !$this->getRequest()->attributes->has('_controller_reflectors')) {
+            $action = new \ReflectionFunction($controller(...));
+            $this->getRequest()->attributes->set('_controller_reflectors', [str_contains($action->name, '{closure}') ? null : $action->getClosureScopeClass(), $action]);
+            unset($this->attributes);
+        }
+
         $this->controller = $controller;
     }
+
+    public function getAttributes(): array
+    {
+        if (isset($this->attributes) || ![$class, $action] = $this->getRequest()->attributes->get('_controller_reflectors')) {
+            return $this->attributes ??= [];
+        }
+
+        foreach (array_merge($class?->getAttributes() ?? [], $action->getAttributes()) as $attribute) {
+            if (class_exists($attribute->getName())) {
+                $this->attributes[$attribute->getName()][] = $attribute->newInstance();
+            }
+        }
+
+        return $this->attributes ??= [];
+    }
+
+    public function setAttributes(array $attributes): void
```

Comment:
should we have a `addAttribute` method instead, that allows adding additional attributes without loosing the ones coming from Reflection ?

Reply:
This works already if one wants to add an attribute without overriding the existing ones:
`$event->setAttributes(array_merge_recursive('my_attr' => [new Foo()], $event->getAttributes())`

So I would reformulate your question as: is this common enough to deserve a simpler API? I'm not sure personally. WDYT?

Reply:
well, the need to perform a recursive merge is easy to overlook. And I don't see any use case where you would want to *remove* the attributes coming from Reflection. This would be a big WTF for someone having to debug why their attributes have no effect.